crates/wasi/src/preview2/ip_name_lookup.rs

```diff
@@ -5,35 +5,72 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
 use crate::preview2::{spawn_blocking, AbortOnDropJoinHandle, SocketError, WasiView};
 use anyhow::Result;
 use std::mem;
-use std::net::{Ipv6Addr, ToSocketAddrs};
+use std::net::{IpAddr, Ipv6Addr, ToSocketAddrs};
 use std::pin::Pin;
 use std::str::FromStr;
 use std::vec;
 use wasmtime::component::Resource;
 
-use super::network::{from_ipv4_addr, from_ipv6_addr};
-
 pub enum ResolveAddressStream {
-    Waiting(AbortOnDropJoinHandle<Result<Vec<IpAddress>, SocketError>>),
-    Done(Result<vec::IntoIter<IpAddress>, SocketError>),
+    Waiting(AbortOnDropJoinHandle<std::io::Result<Vec<IpAddr>>>),
+    Done(std::io::Result<vec::IntoIter<IpAddr>>),
+}
+
+/// A trait for providing a `IpNameLookup` implementation.
+pub trait WasiIpNameLookupView: WasiView {
+    /// The type that implements `IpNameLookup`.
+    type IpNameLookup: IpNameLookup + Send + 'static;
+
+    /// Get a new instance of the name lookup.
+    fn ip_name_lookup(&self) -> Self::IpNameLookup;
+}
+
+/// A trait for resolving IP addresses from a name.
+#[async_trait::async_trait]
+pub trait IpNameLookup {
+    /// Given a name, resolve to a list of IP addresses
+    async fn resolve_addresses(&mut self, name: String) -> std::io::Result<Vec<IpAddr>>;
```

Comment:
I'm curious; why isn't `resolve_addresses` placed directly  on `WasiIpNameLookupView`? Is there something the extra indirection buys us?

Reply:
I'm not sure it gives a ton given that we initialize the type that implements `IpNameLookup` right before we call `resolve_addresses`. The reason I initially went this direction is that it mirrors how the UDP and TCP implementations will work. I can remove it though. Shall I?

Reply:
Alright. Was just curious.
No strong preference